packages/rum-core/src/common/compress.js

```diff
@@ -0,0 +1,294 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { NAVIGATION_TIMING_MARKS } from '../performance-monitoring/capture-navigation'
+
+/**
+ * Compression of all the below schema is based on the v3 RUM Specification
+ * Mapping of existing fields can  be found here
+ * https://github.com/elastic/apm-server/blob/master/model/modeldecoder/field/rum_v3_mapping.go
+ */
+
+const COMPRESSED_NAV_TIMING_MARKS = [
+  'fs',
+  'ls',
+  'le',
+  'cs',
+  'ce',
+  'qs',
+  'rs',
+  're',
+  'dl',
+  'di',
+  'ds',
+  'de',
+  'dc',
+  'es',
+  'ee'
+]
+
+function compressStackFrames(frames) {
+  return frames.map(frame => ({
+    ap: frame.abs_path,
+    f: frame.filename,
+    fn: frame.function,
+    li: frame.lineno,
+    co: frame.colno
+  }))
+}
+
+function compressResponse(response) {
+  return {
+    ts: response.transfer_size,
+    ebs: response.encoded_body_size,
+    dbs: response.decoded_body_size
+  }
+}
+
+function compressHTTP(http) {
+  const compressed = {}
+  for (const key of Object.keys(http)) {
+    const value = http[key]
+    switch (key) {
+      case 'method':
+        compressed.mt = value
+        break
+      case 'status_code':
+        compressed.sc = value
+        break
+      case 'url':
+        compressed.url = value
+        break
+      case 'response':
+        compressed.r = compressResponse(value)
+        break
+    }
+  }
+  return compressed
+}
+
+function compressContext(context) {
+  if (!context) {
+    return null
+  }
+  const compressed = {}
+  for (const key of Object.keys(context)) {
+    const value = context[key]
+    switch (key) {
+      case 'page':
+        compressed.p = {
+          rf: value.referer,
+          url: value.url
+        }
+        break
+      case 'http':
+        compressed.h = compressHTTP(value)
+        break
+      case 'response':
+        compressed.r = compressResponse(value)
+        break
+      case 'user':
+        compressed.u = {
+          id: value.id,
+          un: value.username,
+          em: value.email
+        }
+        break
+      case 'destination':
+        const { service } = value
+        compressed.dt = {
+          se: {
+            n: service.name,
+            t: service.type,
+            rc: service.resource
+          },
+          ad: value.address,
+          po: value.port
+        }
+        break
+      case 'custom':
+        compressed.cu = value
+        break
+    }
+  }
+  return compressed
+}
+
+function compressMarks(marks) {
+  if (!marks) {
+    return null
+  }
+  const { navigationTiming, agent } = marks
+  const compressed = { nt: {}, a: {} }
+
+  COMPRESSED_NAV_TIMING_MARKS.forEach((mark, index) => {
+    const mapping = NAVIGATION_TIMING_MARKS[index]
```

Comment:
Would you please move these two constants in the same file and add a comment that changing the order of one of them would affect the compression?

But a better solution would be to change `NAVIGATION_TIMING_MARKS` to be an object that contains the mapping and remove `COMPRESSED_NAV_TIMING_MARKS`

Reply:
Mapping of these two are not done as we cannot determine whether we have to use key or value depending on whether compression is supported or not at this part of the code. I will move them to same place and add comments.